### PR TITLE
Optimize dockerfile to generate small image

### PR DESCRIPTION
**Pull request by nmaiti**
-> Changed base version to alpine
    -> Multistage docker build to create small deployment image

Signed-off-by: Nabendu Maiti <1206581+nmaiti@users.noreply.github.com>

Docker image size is removed from 1.2Gb to ~70MB

Before Creating a pull request, please read the contributing guidelines thoroughly.

I have fully read and understood the terms and conditions laid down in the general [Contributor License Agreement](https://aahnik.github.io/aahnik/CLA.html)

- [x] I agree to distribute my code contributions under MIT License, and will not change it in the future.
- [x] I agree that any contribution once merged, cannot be taken back by me.
- [x] I will abide by the Code of Conduct
- [x] I understand that the decision of the maintainer is final and abiding. And the maintainer reserves all rights to modify my code. I also understand that the maintainer can remove my code in future, if he thinks so.
- [x] Once my contribution is merged, my name will permanently appear in the Contribtor's List of this repository.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,15 +1,37 @@
...
 
-RUN pip install --upgrade pip wheel setuptools poetry
+ENV PIP_DEFAULT_TIMEOUT=100 \
+    PIP_DISABLE_PIP_VERSION_CHECK=1 \
+    PIP_NO_CACHE_DIR=1 \
+    POETRY_VERSION=1.0.5
 
-ENV POETRY_VIRTUALENVS_IN_PROJECT true
+ENV CRYPTOGRAPHY_DONT_BUILD_RUST=1
+RUN apk add --no-cache gcc libffi-dev musl-dev postgresql-dev postgresql-dev musl-dev zlib zlib-dev jpeg-dev
```
Comment (aahnik) [inline Dockerfile:21]: The thought of using alpine came to my mind earlier, but `tgcf` did not successfully run on alpine. Due to these missing dependencies. 

Manually maintaining these deps in Dockerfile is a hard job for me. Some missing deps can affect some parts of the code. I am afraid that something will break.

In the future, I will be also adding the watermark and OCR plugins, in the default package. So in the Dockerfile, I need to add FFmpeg and tesseract. Installing them in alpine may be a hard job.

The size of the image does not affect run time performance, so I am less interested in optimizing it. 

Most of my users, run tgcf on their computer, installed via pip. And many deploy to Heroku. 

Using alpine may introduce hard to debug problems in the future. I don't have a 100% test coverage, that will allow me to reliably test this image.

So, my suggestion, is that rename this Dockerfile, to `Dockerfile.small`. I will publish this as a separate tag, `tgcf:small`. So this change will be "opt-in" for the end users.

After a sufficient amount of testing (3 to 4 months, by multiple users), we may make this Dockerfile the default.

File: Dockerfile [hunk 2]
```diff
@@ -1,15 +1,37 @@
...
+RUN pip install "poetry==$POETRY_VERSION"
+RUN python -m venv /venv
+
+COPY pyproject.toml poetry.lock ./
+RUN poetry export -f requirements.txt | /venv/bin/pip install -r /dev/stdin
```
Comment (aahnik) [inline Dockerfile:26]: Can you please explain why is this required?

`poetry build` itself is sufficient, I think...

File: Dockerfile (hunk 1 above)
Comment (nmaiti) [reply Dockerfile:21]: oh yes, I initially developed smaller one for my local repo. I will push that way.

> Manually maintaining these deps in Dockerfile is a hard job for me. Some missing deps can affect some parts of the code. I am afraid that something will break.
true but little effort but onetime.

 also pushed https://hub.docker.com/repository/docker/nbmaiti/tgcf for multi platforms.

File: Dockerfile (hunk 2 above)
Comment (nmaiti) [reply Dockerfile:26]: not fully aware about poetry , but what its doing installing poetry dependend package in virtualenv. which may be not reinstalled during poetry build. Will check

Comment (nmaiti) [reply Dockerfile]: yes --not required

File: Dockerfile (hunk 1 above)
Comment (nmaiti) [reply Dockerfile]: Pushed new patch.

Please add the build for minimal in your wiki relevant place

docker build . -f Dockerfile.minimal -t imagename

**Comment by aahnik** [comment]
@nmaiti . this pr is really great. Thanks a lot! 

Just a few final things.

1. Update the Makefile's docker target. Add a new docker build line and a tag line. tag this image with `minimal-$(VERSION)` and `minimal`  (so that when I run `make docker` the minimal image will be built and distributed along with the default image)
2. Click on the conditions checkbox the pr body. (just funny formality)

I will test this docker image, and merge it.

**Comment by aahnik** [comment]
Thanks a lot, @nmaiti for your contribution. In the future, I will try to make the minimal the default.

I have updated the wiki page https://github.com/aahnik/tgcf/wiki/Install-and-run-using-docker

image

------

You may now safely delete your fork to clean up. I see you have a lot of forks in your account. Why don't you clean up?

------

I am curious to connect with you. I did not find any bio/email/social profile. If you have time, please send me a short message via [email](mailto:daw@aahnik.dev) or [telegram](https://telegram.me/aahnikdaw), I have some questions to ask you.
